p2p/discovery/discovery.go

```diff
@@ -0,0 +1,124 @@
+	libp2p_dht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+// Discovery is the interface for the underlying peer discovery protocol.
+// The interface is implemented by dhtDiscovery
+type Discovery interface {
+	Start() error
+	Close() error
+	Advertise(ctx context.Context, ns string) (time.Duration, error)
+	FindPeers(ctx context.Context, ns string, peerLimit int) (<-chan libp2p_peer.AddrInfo, error)
+	GetRawDiscovery() discovery.Discovery
+}
+
+// dhtDiscovery is a wrapper of libp2p dht discovery service. It implements Discovery
+// interface.
+type dhtDiscovery struct {
+	dht  *libp2p_dht.IpfsDHT
+	disc discovery.Discovery
+	host libp2p_host.Host
+
+	opt    DHTConfig
+	logger zerolog.Logger
+	ctx    context.Context
+	cancel func()
+}
+
+// NewDHTDiscovery creates a new dhtDiscovery that implements Discovery interface.
+func NewDHTDiscovery(host libp2p_host.Host, opt DHTConfig) (Discovery, error) {
+	opts, err := opt.getLibp2pRawOptions()
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	dht, err := libp2p_dht.New(ctx, host, opts...)
+	if err != nil {
+		return nil, err
+	}
+	d := libp2p_dis.NewRoutingDiscovery(dht)
+
+	logger := utils.Logger().With().Str("module", "discovery").Logger()
+	return &dhtDiscovery{
+		dht:    dht,
+		disc:   d,
+		host:   host,
+		opt:    opt,
+		logger: logger,
+		ctx:    ctx,
+		cancel: cancel,
+	}, nil
+}
+
+// Start bootstrap the dht discovery service.
+func (d *dhtDiscovery) Start() error {
+	if err := d.connectBootStrapPeers(); err != nil {
+		return err
+	}
+	return d.dht.Bootstrap(d.ctx)
+}
+
+// Stop stop the dhtDiscovery service
+func (d *dhtDiscovery) Close() error {
+	d.cancel()
+	return nil
+}
+
+// Advertise advertises a service
+func (d *dhtDiscovery) Advertise(ctx context.Context, ns string) (time.Duration, error) {
+	return d.disc.Advertise(ctx, ns)
+}
+
+// FindPeers discovers peers providing a service
+func (d *dhtDiscovery) FindPeers(ctx context.Context, ns string, peerLimit int) (<-chan libp2p_peer.AddrInfo, error) {
+	opt := discovery.Limit(peerLimit)
+	return d.disc.FindPeers(ctx, ns, opt)
+}
+
+// GetRawDiscovery get the raw discovery to be used for libp2p pubsub options
+func (d *dhtDiscovery) GetRawDiscovery() discovery.Discovery {
+	return d.disc
+}
+
+// maxNBoostrappers is the maximum bootstrap peers to connect initially.
+// github.com/libp2p/go-libp2p-kad-dht@v0.8.3/dht.go:428
+const maxNBoostrappers = 2
+
+// connectBootStrapPeers connect to bootstrap peers.
+// Currently, specifying bootstrap nodes in dht options does not connect to boot nodes
+// immediately (only after 2 min initial wait in fixLowPeersRoutine). Thus we need to
+// connect to bootstrap peers in this module. Hopefully this logic will be implemented
+// in libp2p naturally.
+func (d *dhtDiscovery) connectBootStrapPeers() error {
+	var found int
+
+	for _, i := range rand.Perm(len(d.opt.BootNodes)) {
```

Comment:
did you test the DNS based boot nodes?

Reply:
Yes. The boot nodes are DNS based in harmony argument, and DNS is resolved in `main.go`

```
// cmd/harmony/main.go:548
bootNodes, err := p2ptypes.ResolveAndParseMultiAddrs(hc.Network.BootNodes)
if err != nil {
    return nil, errors.Wrap(err, "failed to parse boot nodes")
}

myHost, err = p2p.NewHost(p2p.HostConfig{
Self:          &selfPeer,
BLSKey:        nodeConfig.P2PPriKey,
BootNodes:     bootNodes,
DataStoreFile: hc.P2P.DHTDataStore,
})
```

Reply:
Moved the logic from main.go to discovery module since it makes better sense. Thanks.